### ISSUE TITLE
p5.js version

### ISSUE THREAD

**Issue by shiffman** [processing/p5.js-web-editor#144]
@catarak @lmccart and I discussed this on a call yesterday.  At some point in the future we'd like to have a library management system (#4), but for now we need a simple way to keep the p5.js version referenced in `index.html` up-to-date.   We would not change `index.html` in existing projects (or "duplicated" projects) but would simply use the current `p5.js` version (0.5.4 at the time of filing this issue) whenever a "new" project is created.  

@lmccart can you remind us what the url is to check the current p5  version?  It could be checked something like once a day or each time "new" is clicked. 

Did I get this right?

**Comment by lmccart** [comment]
it is here right now: https://github.com/processing/p5.js-website/blob/master/src/data/download.json
I don't think this is the right spot...I'm in the middle of some website restructuring.
but you could use this link for now and I can submit a PR with the updated one if it moves.

**Comment by catarak** [comment]
I noticed that the version file is now at https://github.com/processing/p5.js-website/blob/master/dist/download/version.json, though it's not up to date. Are you just updating that file manually? Would it make more sense to do something like what the website does (https://github.com/processing/p5.js-website/blob/79996764cbff284ae9e8a589065c8cae35949d0b/dist/download/release.php#L30)?

**Comment by lmccart** [comment]
here's the up to date file to link to:
http://p5js.org/download/version.json

it is updated automatically, I think something just got messed up with the git management where an old version of it got cached in the github repo. thanks for catching it!

**Comment by catarak** [comment]
Maybe it is just me but I am getting a 404 error for that link.

**Comment by lmccart** [comment]
ugh I tried to fix the github and deleted the file! should be back now. sorry 😁

**Comment by catarak** [comment]
no worries!

**Comment by dhruvdutt** [comment]
When new project button clicked, we check `"version"` from [here](https://p5js.org/download/version.json) and update p5.js version referenced in `index.html`

@catarak Does this plan sound good?

**Comment by catarak** [comment]
makes sense to me! because of CORS this might need to be added to the API and then the API makes the request. i think this also needs to happen when a user navigates to `/` (which is the same as creating a new sketch, right?)

**Comment by dhruvdutt** [comment]
> p5.js version referenced in index.html

@catarak Sorry, I'm a bit confused. Which `index.html` are you guys referring here? Is it the root `index.html`? I don't see any p5.js version already referenced as per the original post in the issue.

**Comment by catarak** [comment]
the one included in each sketch—if you open the sidebar you can view it!

**Comment by dhruvdutt** [comment]
Oh damn! I always miss this mysteriously hidden menu. :sweat_smile:

**Comment by catarak** [comment]
the ui isn't super clear right now—see #4 for a bit of discussion about this, and #389.

**Comment by catarak** [comment]
i think i actually have figured out a solution for this issue! the p5.js version could be an environment variable (storied in the Kubernetes secret) which gets set via a cron job which checks the version url, once a day probably. that sounds like i said a lot of buzzwords but it is a real solution.

**Comment by jywarren** [comment]
😄 👍 Looking for latest `p5.sound.min.js` for an upcoming workshop, to get the latest fix in https://github.com/processing/p5.js-sound/pull/322 (current is `/*! p5.sound.min.js v0.3.7 2018-01-19 */`)

So, linking to one of these from `index.html` for now: https://cdnjs.com/libraries/p5.js

But also would this be something we could just auto-generate or substitute into one of these sections?

https://github.com/processing/p5.js-web-editor/search?q=cdnjs.cloudflare.com&unscoped_q=cdnjs.cloudflare.com

https://github.com/processing/p5.js-web-editor/blob/5fdcd08adda6cea3408c6331b6ec7f06fd58aa66/client/modules/IDE/reducers/files.js#L12-L18

maybe this section. One thing I'm thinking though, is that once this file is generated (say in an old sketch), you basically have to update the index.html manually. 

I didn't see an option in the CDN link, but is there any way to point at a js file with https://semver.org-style versioning? Like, ideally:

> https://cdnjs.cloudflare.com/ajax/libs/p5.js/^0.8.0/addons/p5.sound.js

OK, this is just pie in the sky, but thanks everyone!

**Comment by catarak** [comment]
@jywarren semver version would be so cool! i think we are on the same page—i would like to write a script that updates an env variable, which would insert the latest p5 version into the snippet of code you put above. i don't think it should update existing sketches, to prevent breaking changes. maybe, with future library management system, there could be a GUI for users to update their p5 version for a sketch.

**Comment by jywarren** [comment]
Cool! Yes. This seems like a good way forward. I could imagine a system too
where we could use a replace-able $VARIABLE in the saved index.html, but
it'd be fragile and also not at all how the web works (perhaps slightly
sadly... HTML variables anyone?). Or... a way to pattern match and replace
the full CDN paths.. but again, fragile and once it breaks it's very hard
for people to untangle. Better avoid magic solutions, i guess.

Thanks!
